server/publications/collections/products.js

```diff
@@ -2,12 +2,49 @@ import _ from "lodash";
 import { Meteor } from "meteor/meteor";
 import { check, Match } from "meteor/check";
 import { SimpleSchema } from "meteor/aldeed:simple-schema";
-import { Products, Shops, Revisions } from "/lib/collections";
+import { Products, Shops, Media, Revisions } from "/lib/collections";
 import { Reaction, Logger } from "/server/api";
 import { RevisionApi } from "/imports/plugins/core/revisions/lib/api/revisions";
-import { findProductMedia } from "./product";
 import { registerSchema } from "@reactioncommerce/reaction-collections";
 
+/**
+ * Helper method that creates a Media cursor for the Products publication. This method returns only the media that has
+ * been set to show up in the grid via `toGrid: 1`
+ * @method findProductsMedia
+ * @param {object} publicationInstance instance of the publication that invokes this method
+ * @param {array} productIds array of productIds
+ * @return {object} Media Cursor containing the product media that matches the selector
+ */
+export function findProductsMedia(publicationInstance, productIds) {
+  const selector = {};
+
+  // Find media that matches the productIds provided in the args
+  selector["metadata.productId"] = {
+    $in: productIds
+  };
+
+  // Find media that is set to show in the grid
+  selector["metadata.toGrid"] = 1;
+
```

Comment:
Not sure how this is determined or what say the UI has in it.

I've seen `toGrid` used to determine this in other places as well, so that's what I defaulted to, but when chatting with @aaronjudd about it, he doesn't remember using `toGrid` so it may be a recent introduction.

Part of the issue here may be that we're no longer publishing every image associated with a product and it's variants to the grid with this PR. Very open to suggestions here.

As far as the image you're seeing on the grid, that seems like an issue with the way that `toGrid` works (or doesn't work), or that for some reason the last product published is getting priority _and_ getting published `toGrid`, or that every variant has a `toGrid` option which might make sense with @mikemurray's note above that in some layouts we display multiple images for a single product. 

I recall a conversation recently where it seemed difficult to set the primary product image because there is always a selected variant, is that still the case and could that be a potential cause here as well?

Reply:
As you can see from the gif, that image is def the primary product. When a product has ONLY the primary product the correct image shows up so, like you said, it seems like somehow it's just deciding the last image is the primary image.

Reply:
So actually I remember being a little confused by this. Now I remember that no images are really attached to the "top" product, so in this case these images that show at the top are attached to the "top-level" variant and then the other ones are attached to the options. So I am not sure exactly how that logic should work, should it just parse down the tree and show the image at the highest level of ancestry? It looks like this just current queries for the top-level `metadata.productId` which would be all images at all levels which is why it's getting the last variant image.

Reply:
To the question is why the image logic works this way:  the original intent on the PDP was not that the default variant was not selected for admin users, and that you could upload images to the default product-> these would be the images you knew would show in the grid, as well the default images to fall back to if the variants/options didn't have any unique images.   At some point, we implemented the first variant being auto selected for all users (which may be a choice that needs review again), but this probably wasn't appropriate for the admin user, and led to a altered pattern here.

Reply:
re: I don't remember `toGrid` and product data doesn't seem to be the appropriate place for layout information to be stored (do we have an existing pattern like, that, if so, we should review).

Reply:
After a bit of a deep dive when creating a product in Reaction, it seems that `toGrid: 1` is set for the first image that is uploaded for a product, as well as the first image of each variant that is uploaded. I've adjusted the find for additionalProductMedia to account for this.

Additionally, it appears that `toGrid` came in from newsiberian about two years ago, probably during the flattening of the products collection.

I think that we probably need to revisit exactly how images are stored and how they are connected to the product grid when we redesign the media gallery.

@zenweasel I'm investigating your report about the wrong media loading right now.

Reply:
@zenweasel I'm not able to reproduce the behavior you're showing at this point. I have changed a few things, but not sure what would have caused that behavior in the first place.
![](http://g.recordit.co/tb4EqlKu3B.gif)